ksqldb-engine/src/main/java/io/confluent/ksql/util/SharedKafkaStreamsRuntimeImpl.java

```diff
@@ -0,0 +1,225 @@
+    } catch (final Exception classificationException) {
+      LOG.error("Error classifying unhandled exception", classificationException);
+    } finally {
+      // If error classification throws then we consider the error to be an UNKNOWN error.
+      // We notify listeners and add the error to the errors queue in the finally block to ensure
+      // all listeners and consumers of the error queue (e.g. the API) can see the error. Similarly,
+      // log in finally block to make sure that if there's ever an error in the classification
+      // we still get this in our logs.
+      final QueryError queryError =
+          new QueryError(
+              System.currentTimeMillis(),
+              Throwables.getStackTraceAsString(e),
+              errorType
+          );
+      for (PersistentQueriesInSharedRuntimesImpl query: metadata.values()) {
+        query.getListener().onError(metadata.get(query.getQueryId().toString()), queryError);
+      }
+      queryErrors.add(queryError);
+      LOG.error(
+          "Unhandled exception caught in streams thread {}. ({})",
+          Thread.currentThread().getName(),
+          errorType,
+          e
+      );
+    }
+    return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.REPLACE_THREAD;
+  }
+
+  public KafkaStreams getKafkaStreams() {
+    return kafkaStreams;
+  }
+
+  public KafkaStreams.State state() {
+    return kafkaStreams.state();
+  }
+
+  public Collection<StreamsMetadata> allMetadata() {
+    return kafkaStreams.metadataForAllStreamsClients();
+  }
+
+  public Set<StreamsTaskMetadata> getTaskMetadata() {
+    return kafkaStreams.metadataForLocalThreads()
+                       .stream()
+                       .flatMap(t -> t.activeTasks().stream())
+                       .map(StreamsTaskMetadata::fromStreamsTaskMetadata)
+                       .collect(Collectors.toSet());
+  }
+
+  public void restart(final QueryId queryId) {
+    final KafkaStreamsNamedTopologyWrapper newKafkaStreams = kafkaStreamsBuilder
+            .build(streamsProperties);
+    newKafkaStreams.setUncaughtExceptionHandler(this::uncaughtHandler);
+    kafkaStreams.close();
+    newKafkaStreams.start();
+    for (PersistentQueriesInSharedRuntimesImpl query: metadata.values()) {
+      newKafkaStreams.addNamedTopology(query.getTopology());
+    }
+    kafkaStreams = newKafkaStreams;
+  }
+
+  public boolean isError(final QueryId queryId) {
+    return !queryErrors.toImmutableList().isEmpty();
+  }
+
+  public void close(final QueryId queryId) {
+    metadata.remove(queryId.toString());
+    if (kafkaStreams.state().isRunningOrRebalancing()) {
+      try {
+        kafkaStreams.removeNamedTopology(queryId.toString());
+      } catch (IllegalArgumentException e) {
+        //don't block
+      }
+    } else {
+      throw new IllegalStateException("Streams in not running but is in state"
+          + kafkaStreams.state());
+    }
+    kafkaStreams.cleanUpNamedTopology(queryId.toString());
```

Comment:
This function would check `getTopologyByName(name).isPresent()` which would return nothing since we already `metadata.remove(queryId.toString());` beforehand, right?

Reply:
the `metadata.remove(queryId.toString())` only affects the ksql metadata I have stored in this runtime class.